doc/ircd.8
.\" @(#)ircd.8 7b10 25 Oct 2001
.\" $Id: ircd.8 1805 2013-04-01 12:01:37Z michael $
.TH IRCD 8 "1 April 2013" "ircd-hybrid-8"
.SH NAME
ircd \- The IRCD-Hybrid Internet Relay Chat server
.SH SYNOPSIS
.hy 0
.IP \fBircd\fP
[-configfile filename] [-dlinefile filename] [-klinefile filename]
[-xlinefile filename] [-logfile filename] [-pidfile filename] [-foreground]
[-version]
.SH DESCRIPTION
.LP
\fIircd\fP is the server (daemon) program for the Internet Relay Chat
Program.  The \fIircd\fP is a server in that its function is to "serve"
the client program \fIirc(1)\fP and other compatible programs with
messages and commands.  All commands and user messages are passed
directly to the \fIircd\fP for processing and relaying to other IRC
servers.

This is \fIircd-hybrid\fP, the highly modified (and hopefully improved)
variant of the original ircd program by Jarkko Oikarinen.
.SH OPTIONS
.TP
.B \-configfile filename
Specifies the ircd.conf file to be used for this \fIircd\fP. The option
is used to override the default ircd.conf given at compile time. By default,
this is "etc/ircd.conf" within the prefix you installed the \fIircd\fP in.
.TP
.B \-dlinefile filename
.TP
.B \-klinefile filename
.TP
.B \-xlinefile filename
Specifies the files to be used for D-lines (host bans), K-lines (hostmask bans),
and X-lines (gecos bans), which by default are within the etc/ directory of your
installation prefix as dline.conf, kline.conf, and xline.conf.
.TP
.B \-logfile filename
Specifies an alternative logfile to be used than that specified in defaults.h
.TP
.B \-pidfile filename
Specifies the file used by the \fIircd\fP to store its process ID. The option is
used to override the default ircd.pid given at compile time.
.TP
.B \-foreground
Makes \fIircd\fP run in the foreground
.TP
.B \-version
Makes \fIircd\fP print its version, and exit.
.SH USAGE
If you plan to connect your \fIircd\fP server to an existing IRC network,
you will need to alter your local IRC configuration file (typically named
"ircd.conf") so that it will accept and make connections to other \fIIRC\fP
servers.  This file contains the hostnames, IP addresses, and sometimes
passwords for connections to other ircds around the world.

The example ircd.conf in the etc/ directory documents itself. Read it 
carefully or you may expose a risk on your network simply by having your 
ircd running!
.LP
.SH BOOTING THE SERVER
It is sufficient to type \fBircd\fP at the command line to start 
ircd-hybrid into the background. If you wish to run ircd in the 
foreground (perhaps for debugging purposes), use the \fB-foreground\fP 
parameter.

The ircd-hybrid package in your distribution may ship with an rc-script 
which handles this for you. In Debian, it is \fB/etc/init.d/ircd-hybrid.\fP

.SH EXAMPLE
# ircd -configfile /usr/share/ircd/ircd.conf -logfile /var/log/ircd.log

Starts \fBircd\fP with the config file /usr/share/ircd/ircd.conf and 
with a log file /var/log/ircd.log. It implicitly forks back to the 
console -- to prevent this use -foreground.

Many of the above paths are hardcoded at compile time in the 
\fBinclude/defaults.h\fP file, so it may be convenient to edit these to 
suit your needs before you compile the ircd.
.SH COPYRIGHT
(c) 1988,1989 University of Oulu, Computing Center, Finland,
.LP
(c) 1988,1989 Department of Information Processing Science,
University of Oulu, Finland
.LP
(c) 1988-1991 Jarkko Oikarinen
.LP
(c) 1997-2013 The IRCD-Hybrid project.
.LP
For full COPYRIGHT see the LICENSE file within the IRC source package.
.LP
.RE
.SH BUGS
If you find a bug and you have a core file from ircd as a result, use a 
debugger like \fBgdb(1)\fP to process the core file and send a backtrace 
to \fBbugs@ircd-hybrid.org\fP.

.SH AUTHORS
irc2.8 and earlier: Jarkko Oikarinen, currently jto@tolsun.oulu.fi. See 
doc/old/Authors for more credits.
.LP
ircd-hybrid-8: the IRCD-Hybrid Project, bugs@ircd-hybrid.org.
.LP
Manual page written by Jeff Trim, jtrim@orion.cair.du.edu,
later modified by jto@tolsun.oulu.fi.
.LP
Updated by W. Campbell, wcampbel@botbay.net, Edward Brocklesby, 
ejb@klamath.uucp.leguin.org.uk, and highly revised by Joshua Kwan 
(joshk@triplehelix.org) for the Debian distribution.


NEWS
-- ircd-hybrid-8.1.7 Release Notes
o) Fixed issue with channel mode +n having no functionality at all
o) Fixed SSL certificate fingerprint validation for outgoing server connects
o) Updated several documentation files


-- ircd-hybrid-8.1.6 Release Notes
o) Fixed possible core with empty motd files


-- ircd-hybrid-8.1.5 Release Notes
o) Fixed bug that would prevent servers from linking together
   if connect::aftype isn't set


-- ircd-hybrid-8.1.4 Release Notes
o) Added 'xline' and 'resv' logging types. See doc/reference.conf
   for more information
o) Fixed bug where remote /STATS requests were not rate limited
o) Fixed core with empty auth::spoof entries
o) Increased oper/auth/connect password length limit from 20 to 128
o) Minor fixes to nickflood control code


-- ircd-hybrid-8.1.3 Release Notes
o) Fixed possible core on "STATS z"
o) Revised doc/reference.conf
o) Fixed broken --disable-libgeoip switch


-- ircd-hybrid-8.1.2 Release Notes
o) Added general::cycle_on_host_change configuration option
o) Added general::stats_u_oper_only configuration option
o) Added support for SHA-256 ssl certificate fingerprint based operator{}
   and connect{} blocks.
   In conjunction with Anope 1.9.x IRC-services, clients are now also able
   to automatically identify for their nick with ssl certificate
   fingerprints
o) Added operator::ssl_connection_required configuration option. See
   doc/reference.conf for more information
o) Added usermode +S (client is connected via SSL/TLS). Allows services
   to keep track of what users are connected via SSL, and allows to see
   ssl-status of remote clients in a /whois
o) Fixed a server name leak with server hiding enabled. Reported by
   Adam <adam@anope.org>


-- ircd-hybrid-8.1.1 Release Notes
########################################################################
o) IMPORTANT: moved disable_remote_command configuration directive from
   general{} block to serverhide{} block
########################################################################
o) Minor code cleanups/performance improvements
o) Fixed bug where opers could see LOCOPS messages even if they don't have
   the +l mode set
o) Fixed bug where non-SSL clients could join +S channels on non-SSL servers
o) Implemented motd{} configuration blocks. See doc/reference.conf for more
   information
o) "STATS T" shows configured MOTD files


-- ircd-hybrid-8.1.0 Release Notes
o) Minor code cleanups/performance improvements


-- ircd-hybrid-8.1.0rc1 Release Notes
o) Fixed broken spoofs


-- ircd-hybrid-8.1.0beta5 Release Notes
o) Removed 'remote', and 'global_kill' oper flags, and added 'connect',
   'squit', and 'kill' flags for better fine-tuning instead. Whether or
   not a specific action is allowed on a remote server can be controlled
   by appending the ':remote' flag. For example: 'kill' allows only local
   clients to be killed, whereas 'kill:remote' allows to issue a KILL for
   remote clients
o) Added 'locops' and 'wallops' to irc-operator flags
o) Improve/cleanup HELP system


-- ircd-hybrid-8.1.0beta4 Release Notes
o) Implemented channel mode +M. Clients that haven't identified their
   name with NickServ may not speak in a channel that has this mode set
o) Fixed weird idletimes shown in /trace
o) Added 'nononreg' (+R) to general::oper_umodes
o) Added user mode +F (can see remote client connect/exit notices)


-- ircd-hybrid-8.1.0beta3 Release Notes
o) PCRE support has been dropped
o) "STATS o" now shows how many times an oper{} block has been used.
   Similar to STATS x|q"
o) Implemented channel mode +c. Known from other ircds, this mode basically
   prevents users from sending messages including control codes to a channel
   that has this mode set
o) Fixed bug where bans were not checked against non-channel members when
   sending messages to a channel
o) Removed channel::quiet_on_ban configuration option. This feature is
   now enabled by default


-- ircd-hybrid-8.1.0beta2 Release Notes
o) Fixed broken compile with libGeoIP disabled
o) Code cleanups; working towards stabilization and improved performance
o) Removed operflag 'nick_changes'. Operators can now set +n at will
o) Fixed shared{} blocks not working as expected
o) Fixed spoofs not working as expected


-- ircd-hybrid-8.1.0beta1 Release Notes
########################################################################
o) IMPORTANT: name/channel entries can't be stacked any longer within
   a single resv{} block. Each entry now requires its own resv{} block.
   Read doc/reference.conf for more details
#######################################################################
o) Added resv::exempt configuration option. Exempt can be either a
   ISO 3166 alpha-2 two letter country code, or a nick!user@host mask.
   CIDR is supported
o) Removed channel::restrict_channels configuration option
o) Preliminary libGeoIP support. Currently only used for exempt entries
   in resv{} blocks
o) Improved WEBIRC authentication; added 'webirc' to auth::flags.
   A "webirc." spoof is now no longer required
o) Implemented new memory pool allocator which basically is based upon Tor's
   mempool allocator for Tor cells
o) Major code cleanups
o) Implemented new binary database storage for X-,D-,K-,G-Lines and RESVs.
   Temporary bans are now stored as well and will persist after a reboot
o) Channel based resv{} blocks may now contain wildcards
o) NICK/JOIN now shows the actual reason of reserved nick-/channelnames
o) contrib/ and its content has been removed from the tree
o) Added serverhide::hide_services configuration option
o) Added 'nononreg' (+R) to oper::umodes and general::oper_only_modes
o) Added support for "away-notify" client capability


-- ircd-hybrid-8.0.9 Release Notes
o) Fixed bug where ircd would sometimes drop a services link because
   of a missing argument to the SVSMODE command
o) Fixed weird idletimes shown in /trace


-- ircd-hybrid-8.0.8 Release Notes
o) "STATS s" now shows configured services{} blocks as well

src/ircd.c
/*
 *  ircd-hybrid: an advanced Internet Relay Chat Daemon(ircd).
 *  ircd.c: Starts up and runs the ircd.
 *
 *  Copyright (C) 2002 by the past and present ircd coders, and others.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 *  USA
 *
 *  $Id: ircd.c 2329 2013-06-23 13:29:37Z michael $
 */

#include "stdinc.h"
#include "s_user.h"
#include "list.h"
#include "ircd.h"
#include "channel.h"
#include "channel_mode.h"
#include "client.h"
#include "event.h"
#include "fdlist.h"
#include "hash.h"
#include "irc_string.h"
#include "ircd_signal.h"
#include "s_gline.h"
#include "motd.h"
#include "conf.h"
#include "hostmask.h"
#include "numeric.h"
#include "packet.h"
#include "parse.h"
#include "irc_res.h"
#include "restart.h"
#include "rng_mt.h"
#include "s_auth.h"
#include "s_bsd.h"
#include "log.h"
#include "s_misc.h"
#include "s_serv.h"      /* try_connections */
#include "send.h"
#include "whowas.h"
#include "modules.h"
#include "memory.h"
#include "mempool.h"
#include "hook.h"
#include "ircd_getopt.h"
#include "supported.h"
#include "watch.h"
#include "conf_db.h"
#include "conf_class.h"


#ifdef HAVE_LIBGEOIP
GeoIP *geoip_ctx;
#endif
/* /quote set variables */
struct SetOptions GlobalSetOptions;

/* configuration set from ircd.conf */
struct config_file_entry ConfigFileEntry; 
/* server info set from ircd.conf */
struct server_info ServerInfo;
/* admin info set from ircd.conf */
struct admin_info AdminInfo = { NULL, NULL, NULL };
struct Counter Count = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
struct ServerState_t server_state = { 0 };
struct logging_entry ConfigLoggingEntry = { .use_logging = 1 };
struct ServerStatistics ServerStats;
struct timeval SystemTime;
struct Client me;             /* That's me */
struct LocalUser meLocalUser; /* That's also part of me */

const char *logFileName = LPATH;
const char *pidFileName = PPATH;

char **myargv;

int dorehash = 0;
int doremotd = 0;

/* Set to zero because it should be initialized later using
 * initialize_server_capabs
 */
int default_server_capabs = 0;
unsigned int splitmode;
unsigned int splitchecking;
unsigned int split_users;
unsigned int split_servers;

/* Do klines the same way hybrid-6 did them, i.e. at the
 * top of the next io_loop instead of in the same loop as
 * the klines are being applied.
 *
 * This should fix strange CPU starvation as very indirectly reported.
 * (Why do you people not email bug reports? WHY? WHY?)
 *
 * - Dianora
 */

int rehashed_klines = 0;


/*
 * print_startup - print startup information
 */
static void
print_startup(int pid)
{
  printf("ircd: version %s\n", ircd_version);
  printf("ircd: pid %d\n", pid);
  printf("ircd: running in %s mode from %s\n", !server_state.foreground ? "background"
         : "foreground", ConfigFileEntry.dpath);
}

static void
make_daemon(void)
{
  int pid;

  if ((pid = fork()) < 0)
  {
    perror("fork");
    exit(EXIT_FAILURE);
  }
  else if (pid > 0)
  {
    print_startup(pid);
    exit(EXIT_SUCCESS);
  }

  setsid();
}

static int printVersion = 0;

static struct lgetopt myopts[] = {
  {"configfile", &ConfigFileEntry.configfile, 
   STRING, "File to use for ircd.conf"},
  {"glinefile",  &ConfigFileEntry.glinefile,
   STRING, "File to use for gline database"},
  {"klinefile",  &ConfigFileEntry.klinefile, 
   STRING, "File to use for kline database"},
  {"dlinefile",  &ConfigFileEntry.dlinefile,
   STRING, "File to use for dline database"},
  {"xlinefile",  &ConfigFileEntry.xlinefile, 
   STRING, "File to use for xline database"},
  {"resvfile",  &ConfigFileEntry.resvfile,
   STRING, "File to use for resv database"},
  {"logfile",    &logFileName, 
   STRING, "File to use for ircd.log"},
  {"pidfile",    &pidFileName,
   STRING, "File to use for process ID"},
  {"foreground", &server_state.foreground, 
   YESNO, "Run in foreground (don't detach)"},
  {"version",    &printVersion, 
   YESNO, "Print version and exit"},
  {"help", NULL, USAGE, "Print this text"},
  {NULL, NULL, STRING, NULL},
};

void
set_time(void)
{
  static char to_send[200];
  struct timeval newtime;
  newtime.tv_sec  = 0;
  newtime.tv_usec = 0;

  if (gettimeofday(&newtime, NULL) == -1)
  {
    ilog(LOG_TYPE_IRCD, "Clock Failure (%s), TS can be corrupted",
         strerror(errno));
    sendto_realops_flags(UMODE_ALL, L_ALL, SEND_NOTICE,
                         "Clock Failure (%s), TS can be corrupted",
                         strerror(errno));
    restart("Clock Failure");
  }

  if (newtime.tv_sec < CurrentTime)
  {
    snprintf(to_send, sizeof(to_send),
             "System clock is running backwards - (%lu < %lu)",
             (unsigned long)newtime.tv_sec, (unsigned long)CurrentTime);
    report_error(L_ALL, to_send, me.name, 0);
    set_back_events(CurrentTime - newtime.tv_sec);
  }

  SystemTime.tv_sec  = newtime.tv_sec;
  SystemTime.tv_usec = newtime.tv_usec;
}

static void
io_loop(void)
{
  while (1 == 1)
  {
    /*
     * Maybe we want a flags word?
     * ie. if (REHASHED_KLINES(global_flags)) 
     * SET_REHASHED_KLINES(global_flags)
     * CLEAR_REHASHED_KLINES(global_flags)
     *
     * - Dianora
     */
    if (rehashed_klines)
    {
      check_conf_klines();
      rehashed_klines = 0;
    }

    if (listing_client_list.head)
    {
      dlink_node *ptr = NULL, *ptr_next = NULL;
      DLINK_FOREACH_SAFE(ptr, ptr_next, listing_client_list.head)
      {
        struct Client *client_p = ptr->data;
        assert(client_p->localClient->list_task);
        safe_list_channels(client_p, client_p->localClient->list_task, 0);
      }
    }

    /* Run pending events, then get the number of seconds to the next
     * event
     */
    while (eventNextTime() <= CurrentTime)
      eventRun();

    comm_select();
    exit_aborted_clients();
    free_exited_clients();
    send_queued_all();

    /* Check to see whether we have to rehash the configuration .. */
    if (dorehash)
    {
      rehash(1);
      dorehash = 0;
    }
    if (doremotd)
    {
      motd_recache();
      sendto_realops_flags(UMODE_ALL, L_ALL, SEND_NOTICE,
                           "Got signal SIGUSR1, reloading motd files");
      doremotd = 0;
    }
  }
}

/* initalialize_global_set_options()
 *
 * inputs       - none
 * output       - none
 * side effects - This sets all global set options needed 
 */
static void
initialize_global_set_options(void)
{
  memset(&GlobalSetOptions, 0, sizeof(GlobalSetOptions));

  GlobalSetOptions.autoconn  = 1;
  GlobalSetOptions.spam_time = MIN_JOIN_LEAVE_TIME;
  GlobalSetOptions.spam_num  = MAX_JOIN_LEAVE_COUNT;

  if (ConfigFileEntry.default_floodcount)
    GlobalSetOptions.floodcount = ConfigFileEntry.default_floodcount;
  else
    GlobalSetOptions.floodcount = 10;

  /* XXX I have no idea what to try here - Dianora */
  GlobalSetOptions.joinfloodcount = 16;
  GlobalSetOptions.joinfloodtime = 8;

  split_servers = ConfigChannel.default_split_server_count;
  split_users   = ConfigChannel.default_split_user_count;

  if (split_users && split_servers && (ConfigChannel.no_create_on_split ||
                                       ConfigChannel.no_join_on_split))
  {
    splitmode     = 1;
    splitchecking = 1;
  }

  GlobalSetOptions.ident_timeout = IDENT_TIMEOUT;
  /* End of global set options */
}

/* initialize_server_capabs()
 *
 * inputs       - none
 * output       - none
 */
static void
initialize_server_capabs(void)
{
  add_capability("QS", CAP_QS, 1);
  add_capability("EOB", CAP_EOB, 1);
  add_capability("TS6", CAP_TS6, 0);
  add_capability("CLUSTER", CAP_CLUSTER, 1);
//  add_capability("FAKEHOST", CAP_FAKEHOST, 1);
  add_capability("SVS", CAP_SVS, 1);
#ifdef HALFOPS
  add_capability("HOPS", CAP_HOPS, 1);
#endif
}

/* write_pidfile()
 *
 * inputs       - filename+path of pid file
 * output       - NONE
 * side effects - write the pid of the ircd to filename
 */
static void
write_pidfile(const char *filename)
{
  FILE *fb;

  if ((fb = fopen(filename, "w")))
  {
    char buff[32];
    unsigned int pid = (unsigned int)getpid();

    snprintf(buff, sizeof(buff), "%u\n", pid);

    if ((fputs(buff, fb) == -1))
      ilog(LOG_TYPE_IRCD, "Error writing %u to pid file %s (%s)",
           pid, filename, strerror(errno));

    fclose(fb);
  }
  else
  {
    ilog(LOG_TYPE_IRCD, "Error opening pid file %s", filename);
  }
}

/* check_pidfile()
 *
 * inputs       - filename+path of pid file
 * output       - none
 * side effects - reads pid from pidfile and checks if ircd is in process
 *                list. if it is, gracefully exits
 * -kre
 */
static void
check_pidfile(const char *filename)
{
  FILE *fb;
  char buff[32];
  pid_t pidfromfile;

  /* Don't do logging here, since we don't have log() initialised */
  if ((fb = fopen(filename, "r")))
  {
    if (fgets(buff, 20, fb) == NULL)
    {
      /* log(L_ERROR, "Error reading from pid file %s (%s)", filename,
       * strerror(errno));
       */
    }
    else
    {
      pidfromfile = atoi(buff);

      if (!kill(pidfromfile, 0))
      {
        /* log(L_ERROR, "Server is already running"); */
        printf("ircd: daemon is already running\n");
        exit(-1);
      }
    }

    fclose(fb);
  }
  else if (errno != ENOENT)
  {
    /* log(L_ERROR, "Error opening pid file %s", filename); */
  }
}

/* setup_corefile()
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups corefile to system limits.
 * -kre
 */
static void
setup_corefile(void)
{
#ifdef HAVE_SYS_RESOURCE_H
  struct rlimit rlim; /* resource limits */

  /* Set corefilesize to maximum */
  if (!getrlimit(RLIMIT_CORE, &rlim))
  {
    rlim.rlim_cur = rlim.rlim_max;
    setrlimit(RLIMIT_CORE, &rlim);
  }
#endif
}

#ifdef HAVE_LIBCRYPTO
static int
always_accept_verify_cb(int preverify_ok, X509_STORE_CTX *x509_ctx)
{
  return 1;
}
#endif

/* init_ssl()
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups SSL context.
 */
static void
ssl_init(void)
{
#ifdef HAVE_LIBCRYPTO
  SSL_load_error_strings();
  SSLeay_add_ssl_algorithms();

  if ((ServerInfo.server_ctx = SSL_CTX_new(SSLv23_server_method())) == NULL)
  {
    const char *s;

    fprintf(stderr, "ERROR: Could not initialize the SSL Server context -- %s\n",
            s = ERR_lib_error_string(ERR_get_error()));
    ilog(LOG_TYPE_IRCD, "ERROR: Could not initialize the SSL Server context -- %s\n", s);
  }

  SSL_CTX_set_options(ServerInfo.server_ctx, SSL_OP_NO_SSLv2|SSL_OP_NO_SSLv3|SSL_OP_NO_TLSv1);
  SSL_CTX_set_options(ServerInfo.server_ctx, SSL_OP_TLS_ROLLBACK_BUG|SSL_OP_ALL);
  SSL_CTX_set_verify(ServerInfo.server_ctx, SSL_VERIFY_PEER|SSL_VERIFY_CLIENT_ONCE,
                     always_accept_verify_cb);

  if ((ServerInfo.client_ctx = SSL_CTX_new(SSLv23_client_method())) == NULL)
  {
    const char *s;

    fprintf(stderr, "ERROR: Could not initialize the SSL Client context -- %s\n",
            s = ERR_lib_error_string(ERR_get_error()));
    ilog(LOG_TYPE_IRCD, "ERROR: Could not initialize the SSL Client context -- %s\n", s);
  }

  SSL_CTX_set_options(ServerInfo.client_ctx, SSL_OP_NO_SSLv2|SSL_OP_NO_SSLv3|SSL_OP_NO_TLSv1);
  SSL_CTX_set_options(ServerInfo.client_ctx, SSL_OP_TLS_ROLLBACK_BUG|SSL_OP_ALL);
  SSL_CTX_set_verify(ServerInfo.client_ctx, SSL_VERIFY_PEER|SSL_VERIFY_CLIENT_ONCE,
                     always_accept_verify_cb);
#endif /* HAVE_LIBCRYPTO */
}

int
main(int argc, char *argv[])
{
  /* Check to see if the user is running us as root, which is a nono */
  if (geteuid() == 0)
  {
    fprintf(stderr, "Don't run ircd as root!!!\n");
    return -1;
  }

  /* Setup corefile size immediately after boot -kre */
  setup_corefile();

  /* save server boot time right away, so getrusage works correctly */
  set_time();

  /* It ain't random, but it ought to be a little harder to guess */
  init_genrand(SystemTime.tv_sec ^ (SystemTime.tv_usec | (getpid() << 20)));

  me.localClient = &meLocalUser;
  dlinkAdd(&me, &me.node, &global_client_list);  /* Pointer to beginning
						   of Client list */
  /* Initialise the channel capability usage counts... */
  init_chcap_usage_counts();

  ConfigFileEntry.dpath      = DPATH;
  ConfigFileEntry.configfile = CPATH;    /* Server configuration file */
  ConfigFileEntry.klinefile  = KPATH;    /* Server kline file         */
  ConfigFileEntry.glinefile  = GPATH;    /* Server gline file         */
  ConfigFileEntry.xlinefile  = XPATH;    /* Server xline file         */
  ConfigFileEntry.dlinefile  = DLPATH;   /* dline file                */
  ConfigFileEntry.resvfile   = RESVPATH; /* resv file                 */

  myargv = argv;
  umask(077);                /* better safe than sorry --SRB */

  parseargs(&argc, &argv, myopts);

  if (printVersion)
  {
    printf("ircd: version %s\n", ircd_version);
    exit(EXIT_SUCCESS);
  }

  if (chdir(ConfigFileEntry.dpath))
  {
    perror("chdir");
    exit(EXIT_FAILURE);
  }

  ssl_init();

  if (!server_state.foreground)
  {
    make_daemon();
    close_standard_fds(); /* this needs to be before init_netio()! */
  }
  else
    print_startup(getpid());

  setup_signals();

  /* Init the event subsystem */
  eventInit();

  /* We need this to initialise the fd array before anything else */
  fdlist_init();
  log_set_file(LOG_TYPE_IRCD, 0, logFileName);
  check_can_use_v6();
  init_comm();         /* This needs to be setup early ! -- adrian */

  /* Check if there is pidfile and daemon already running */
  check_pidfile(pidFileName);

  mp_pool_init();
  init_dlink_nodes();
  init_isupport();
  dbuf_init();
  hash_init();
  init_ip_hash_table();      /* client host ip hash table */
  init_host_hash();          /* Host-hashtable. */
  client_init();
  class_init();
  whowas_init();
  watch_init();
  auth_init();          /* Initialise the auth code */
  init_resolver();      /* Needs to be setup before the io loop */
  modules_init();
  read_conf_files(1);   /* cold start init conf files */
  init_uid();
  initialize_server_capabs();   /* Set up default_server_capabs */
  initialize_global_set_options();
  channel_init();
  read_links_file();
  motd_init();
#ifdef HAVE_LIBGEOIP
  geoip_ctx = GeoIP_new(GEOIP_MEMORY_CACHE);
#endif

  if (EmptyString(ServerInfo.sid))
  {
    ilog(LOG_TYPE_IRCD, "ERROR: No server id specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.id, ServerInfo.sid, sizeof(me.id));

  if (EmptyString(ServerInfo.name))
  {
    ilog(LOG_TYPE_IRCD, "ERROR: No server name specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.name, ServerInfo.name, sizeof(me.name));

  /* serverinfo{} description must exist.  If not, error out.*/
  if (EmptyString(ServerInfo.description))
  {
    ilog(LOG_TYPE_IRCD, "ERROR: No server description specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }

  strlcpy(me.info, ServerInfo.description, sizeof(me.info));

  me.from                   = &me;
  me.servptr                = &me;
  me.localClient->lasttime  = CurrentTime;
  me.localClient->since     = CurrentTime;
  me.localClient->firsttime = CurrentTime;

  SetMe(&me);
  make_server(&me);

  hash_add_id(&me);
  hash_add_client(&me);
  
  /* add ourselves to global_serv_list */
  dlinkAdd(&me, make_dlink_node(), &global_serv_list);

  load_kline_database();
  load_dline_database();
  load_gline_database();
  load_xline_database();
  load_resv_database();

  if (chdir(MODPATH))
  {
    ilog(LOG_TYPE_IRCD, "Could not load core modules. Terminating!");
    exit(EXIT_FAILURE);
  }

  load_all_modules(1);
  load_conf_modules();
  load_core_modules(1);

  /* Go back to DPATH after checking to see if we can chdir to MODPATH */
  if (chdir(ConfigFileEntry.dpath))
  {
    perror("chdir");
    exit(EXIT_FAILURE);
  }

  /*
   * assemble_umode_buffer() has to be called after
   * reading conf/loading modules.
   */
  assemble_umode_buffer();

  write_pidfile(pidFileName);

  ilog(LOG_TYPE_IRCD, "Server Ready");

  eventAddIsh("cleanup_glines", cleanup_glines, NULL, CLEANUP_GLINES_TIME);
  eventAddIsh("cleanup_tklines", cleanup_tklines, NULL, CLEANUP_TKLINES_TIME);

  /* We want try_connections to be called as soon as possible now! -- adrian */
  /* No, 'cause after a restart it would cause all sorts of nick collides */
  eventAddIsh("try_connections", try_connections, NULL, STARTUP_CONNECTIONS_TIME);

  /* Setup the timeout check. I'll shift it later :)  -- adrian */
  eventAddIsh("comm_checktimeouts", comm_checktimeouts, NULL, 1);

  eventAddIsh("save_all_databases", save_all_databases, NULL, DATABASE_UPDATE_TIMEOUT);

  if (ConfigServerHide.links_delay > 0)
    eventAddIsh("write_links_file", write_links_file, NULL, ConfigServerHide.links_delay);
  else
    ConfigServerHide.links_disabled = 1;

  if (splitmode)
    eventAddIsh("check_splitmode", check_splitmode, NULL, 60);

  io_loop();
  return 0;
}
